### DISCUSSION TITLE
msg.value in solidity

### DISCUSSION THREAD

**Discussion by Ghadi8** [Cyfrin/assembly-evm-opcodes-and-formal-verification-course#26] [General]
Hey, first of all thanks for the great course!  While deconstructing the bytecodes provided by the solc compiler, we saw the msg.value  being loaded into the stack with CALLVALUE and being duplicated every time. Why does it duplicate it, if it will end up either reverting or popping it after the jumpdest is complete?

**Comment by EngrPips** [comment]
I haven't actually noticed that but maybe that is some of the optimization we can do if we go ahead to write our contract opcodes ourselves but then I doubt the compiler will just unnecessarily duplicate stuff it will never need.
